Add macOS specific fixes and build config for llvm from homebrew

vespakoen opened the pull request:
Just tried compiling it again with a meson cross file pointing to the homebrew llvm installation which seemed to work now, I think is a better alternative to #25.

vespakoen commented:
Looks like in github actions it's not finding the math library now, I will clean my environment and the cache and hopefully get this one to work..

I'll also take a look at the stdlib option with the default clang.

vespakoen commented:
So... I worked around the meson issue (https://github.com/mesonbuild/meson/issues/12622) using some hacks, and they work fine locally (on a arm64 mac) but don't seem to work in github actions some library is missing or it is not compiled with C++20? Anyways, I guess it's a start and might be good enough to be merged at this point.

@carrotIndustries please review ;)

carrotIndustries commented:
The actual code changes look good to me. The thing with extracting the linker invocation from the build logs seems a bit hacky though.

If it's only opencascade with cmake for that's causing it, it may be worth investigating to hardcode its libraries on mac os rather than asking cmake. Like such: https://github.com/horizon-eda/horizon/blob/b78d8063cde1101a56a26562a8206e51e3e45add/meson.build#L31-L35

vespakoen commented:
Thanks for that example! I modified the `meson.build` a bit to make use of that method and added all the opencascade libs manually to the list that I recovered from a previous build log.

I was hoping to get away with the hack and wait for meson to fix the real issue, but since it's an edge case, it might take a while before it gets fixed...

Hope this works well enough...

carrotIndustries commented on meson.build:
```diff
@@ -9,15 +9,20 @@ is_windows = target_machine.system() == 'windows'
 
 
-opencascade = dependency('OpenCASCADE', method : 'cmake')
+opencascade = dependency('OpenCASCADE', method : 'cmake', required:false)
```
maybe make required dependent on `target_machine.system() == 'darwin'` so that it still errors out properly if it's not there on non-mac platforms

carrotIndustries commented on meson.build:
```diff
@@ -9,15 +9,20 @@ is_windows = target_machine.system() == 'windows'
+  message('meson\'s framework handling on macos with the cmake method is broken, replacing')
+  brew_prefix_cmd = run_command('brew', '--prefix', check: true)
+  brew_prefix = brew_prefix_cmd.stdout().strip() + '/lib'
+  opencascade = declare_dependency(dependencies: opencascade.partial_dependency(compile_args: true, includes:true), link_args:['-Wl,-L' + brew_prefix, '-lTKernel', '-lTKMath', '-lTKG2d', '-lTKG3d', '-lTKGeomBase', '-lTKBRep', '-lTKGeomAlgo', '-lTKTopAlgo', '-lTKPrim', '-lTKBO', '-lTKShHealing', '-lTKBool', '-lTKHLR', '-lTKFillet', '-lTKOffset', '-lTKFeat', '-lTKMesh', '-lTKXMesh', '-lTKService',  ...
```
We probably don't need all of these.

vespakoen replied:
Without specifying `'-Wl,-L' + brew_prefix` I get linking errors, because the homebrew path is not searched in by default, alternatively those could be specified with LDFLAGS?

carrotIndustries replied:
Sorry, I was referring referring to the libraries listed on line 24. The ones listed in the Horizon EDA repo should be a good starting point.

vespakoen replied:
Ah I see, I'll try trimming it down

vespakoen replied:
I used the ones from HorizonEDA and added `TKSTL` and `TKFillet` which seemed to do the trick.

vespakoen commented:
And I guess the reason for this error (which shows up in CI):

```
Undefined symbols for architecture x86_64:
  "std::__1::to_chars(char*, char*, double)", referenced from:
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<double>(char*, char*, double) in src_render_renderer.cpp.o
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<double>(char*, char*, double) in src_preferences_preferences_window_canvas.cpp.o
  "std::__1::to_chars(char*, char*, double, std::__1::chars_format)", referenced from:
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<double>(char*, char*, double, std::__1::chars_format) in src_render_renderer.cpp.o
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<double>(char*, char*, double, std::__1::chars_format) in src_preferences_preferences_window_canvas.cpp.o
  "std::__1::to_chars(char*, char*, double, std::__1::chars_format, int)", referenced from:
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<double>(char*, char*, double, std::__1::chars_format, int) in src_render_renderer.cpp.o
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<double>(char*, char*, double, std::__1::chars_format, int) in src_preferences_preferences_window_canvas.cpp.o
  "std::__1::to_chars(char*, char*, long double)", referenced from:
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<long double>(char*, char*, long double) in src_render_renderer.cpp.o
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<long double>(char*, char*, long double) in src_preferences_preferences_window_canvas.cpp.o
  "std::__1::to_chars(char*, char*, long double, std::__1::chars_format)", referenced from:
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<long double>(char*, char*, long double, std::__1::chars_format) in src_render_renderer.cpp.o
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<long double>(char*, char*, long double, std::__1::chars_format) in src_preferences_preferences_window_canvas.cpp.o
  "std::__1::to_chars(char*, char*, long double, std::__1::chars_format, int)", referenced from:
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<long double>(char*, char*, long double, std::__1::chars_format, int) in src_render_renderer.cpp.o
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<long double>(char*, char*, long double, std::__1::chars_format, int) in src_preferences_preferences_window_canvas.cpp.o
  "std::__1::to_chars(char*, char*, float)", referenced from:
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<float>(char*, char*, float) in src_render_renderer.cpp.o
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<float>(char*, char*, float) in src_preferences_preferences_window_canvas.cpp.o
  "std::__1::to_chars(char*, char*, float, std::__1::chars_format)", referenced from:
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<float>(char*, char*, float, std::__1::chars_format) in src_render_renderer.cpp.o
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<float>(char*, char*, float, std::__1::chars_format) in src_preferences_preferences_window_canvas.cpp.o
  "std::__1::to_chars(char*, char*, float, std::__1::chars_format, int)", referenced from:
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<float>(char*, char*, float, std::__1::chars_format, int) in src_render_renderer.cpp.o
      char* std::__1::__formatter::__to_buffer[abi:ue170006]<float>(char*, char*, float, std::__1::chars_format, int) in src_preferences_preferences_window_canvas.cpp.o
ld: symbol(s) not found for architecture x86_64
```

Is simply that `libc++` is not new enough in the macOS github runner image, and is new enough on my macOS version (Ventura 13.5.1).
I am not sure if linking with llvm's libc++ will work well on macOS, perhaps this can be used: https://github.com/fmtlib/fmt ?

vespakoen commented:
In the last commit I dropped the meson cross files and used environment variables to point to homebrew's llvm, I also added the search path for llvm's libc++, which hopefully fixes the std::format issue on older macs.

vespakoen commented:
Yay! that seemed to have worked, I also added some docs to the README and I'll now call it a day, thanks for making `dune3d` it's exactly what I hoped for to happen one day!!

carrotIndustries commented:
That's great news! Does the app itself work, including selecting things?

tkornack commented:
When I built it earlier today it was functional but selection on a high dpi screen seemed to require rather high precision.

vespakoen commented:
The app works indeed, but as mentioned in the last PR, because of the 1 pixel lines, selecting things is hard, and we should probably use quads in stead of lines, especially since the glLineWidth (or whatever it's called) is deprecated.

There are also some Gtk issues for me in combination with my tiling window manager (yabai) that causes dune3d to have padding around the window, and causes it to not to resize, so I can only run it in full screen mode or as a very small (start size) window. I don't know what it looks like without my custom window manager (will try it another day)

I'll go ahead and offer 200€ for the person implementing thick line rendering for macOS and getting it merged :)

carrotIndustries commented:
Will you clean up the commits or should I do a squash merge?

carrotIndustries commented:
> because of the 1 pixel lines, selecting things is hard, and we should probably use quads in stead of lines, especially since the glLineWidth is deprecated.

One doesn't need to hit the lines spot on to be selected, but the fuzz factor probably has to be scaled with the hidpi scaling factor.

For drawing lines using quads, see https://github.com/horizon-eda/horizon/blob/master/src/canvas/shaders/triangle-line0-geometry.glsl

vespakoen commented:
I'm not at a computer, squash merge is fine.

That explains the selection fiddlyness, is also seems to be a couple pixels "off" sometimes, which might be another bug.

Glad that some work for the quad lines is available, but I don't think I have enough experience to integrate it, hoping someone else can and will do it :)

vespakoen commented:
> I'll go ahead and offer 200€ for the person implementing thick line rendering and selection for macOS and getting it merged :)

@carrotIndustries I just re-compiled the latest source and to my delight, saw that the lines are now thick! Can you reach out to me with your paypal or IBAN so I can put my money where my mouth is?

koen [at] schmeets [dot] de

Thanks a lot!
